pkg/store/cache/caching_bucket.go

```diff
@@ -132,35 +131,184 @@ func (cb *CachingBucket) ReaderWithExpectedErrs(expectedFunc objstore.IsOpFailur
+	var list []string
+	err := cb.Bucket.Iter(ctx, dir, func(s string) error {
+		list = append(list, s)
+		return f(s)
+	})
+
+	remainingTTL := cfg.ttl - time.Since(iterTime)
+	if err == nil && remainingTTL > 0 {
+		if data, err := cfg.codec.Encode(list); err == nil {
+			cfg.cache.Store(ctx, map[string][]byte{key: data}, remainingTTL)
+		} else {
+			level.Warn(cb.logger).Log("msg", "failed to encode Iter result", "err", err)
+		}
+	}
+	return err
+}
+
+func (cb *CachingBucket) Exists(ctx context.Context, name string) (bool, error) {
+	cfgName, cfg := cb.cfg.findExistConfig(name)
+	if cfg == nil {
+		return cb.Bucket.Exists(ctx, name)
+	}
+
+	cb.operationRequests.WithLabelValues(opExists, cfgName).Inc()
+
+	key := cachingKeyExists(name)
+	hits := cfg.cache.Fetch(ctx, []string{key})
+
+	if ex := hits[key]; ex != nil {
+		exists, err := strconv.ParseBool(string(ex))
+		if err == nil {
+			cb.operationHits.WithLabelValues(opExists, cfgName).Inc()
+			return exists, nil
+		}
+		level.Warn(cb.logger).Log("msg", "unexpected cached 'exists' value", "val", string(ex))
+	}
+
+	existsTime := time.Now()
+	ok, err := cb.Bucket.Exists(ctx, name)
+	if err == nil {
+		storeExistsCacheEntry(ctx, key, ok, existsTime, cfg.cache, cfg.existsTTL, cfg.doesntExistTTL)
+	}
+
+	return ok, err
+}
+
+func storeExistsCacheEntry(ctx context.Context, cachingKey string, exists bool, ts time.Time, cache cache.Cache, existsTTL, doesntExistTTL time.Duration) {
+	var ttl time.Duration
+	if exists {
+		ttl = existsTTL - time.Since(ts)
+	} else {
+		ttl = doesntExistTTL - time.Since(ts)
+	}
+
+	if ttl > 0 {
+		cache.Store(ctx, map[string][]byte{cachingKey: []byte(strconv.FormatBool(exists))}, ttl)
+	}
+}
+
+func (cb *CachingBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	cfgName, cfg := cb.cfg.findGetConfig(name)
+	if cfg == nil {
+		return cb.Bucket.Get(ctx, name)
+	}
+
+	cb.operationRequests.WithLabelValues(opGet, cfgName).Inc()
+
+	contentKey := cachingKeyContent(name)
+	existsKey := cachingKeyExists(name)
+
+	hits := cfg.cache.Fetch(ctx, []string{contentKey, existsKey})
+	if hits[contentKey] != nil {
+		cb.operationHits.WithLabelValues(opGet, cfgName).Inc()
+		return ioutil.NopCloser(bytes.NewReader(hits[contentKey])), nil
+	}
+
+	// If we know that file doesn't exist, we can return that. Useful for deletion marks.
+	if ex := hits[existsKey]; ex != nil {
+		if exists, err := strconv.ParseBool(string(ex)); err == nil && !exists {
+			cb.operationHits.WithLabelValues(opGet, cfgName).Inc()
+			return nil, errObjNotFound
+		}
+	}
+
+	getTime := time.Now()
+	reader, err := cb.Bucket.Get(ctx, name)
+	if err != nil {
+		if cb.Bucket.IsObjNotFoundErr(err) {
+			// Cache that object doesn't exist.
+			storeExistsCacheEntry(ctx, existsKey, false, getTime, cfg.cache, cfg.existsTTL, cfg.doesntExistTTL)
+		}
+
+		return nil, err
+	}
+	defer runutil.CloseWithLogOnErr(cb.logger, reader, "CachingBucket.Get(%q)", name)
+
+	data, err := ioutil.ReadAll(reader)
```

Comment:
Ok, I think this is wrong. The interface was made to allow you to stream download. We just break it here. Ideally, we want a special readers handling this. This will be complicated and add some work, so I would say, let's leave with this but add proper TODOs and warning comments for this structure. Given it's used for meta jsons it's ok, but it's easy to miss this hidden knowledge (since it's generic wrapper) and use it for huge files.

Reply:
Actually wait a second. You just did all the hard work with cachedGetSubrange!

Let's just use this then? (:

Reply:
Reader used in cachedGetSubrange is different – it fetches all blocks into memory first.

I've now modified `Get` to use streaming, and buffer read object in the memory up to configured size. If object is read fully, we cache it (under single key, unlike `GetRange`, which caches each subrange under separate key).